diff --git a/CHERICC.bsv b/CHERICC.bsv
index 5382f8f..fb993f7 100644
--- a/CHERICC.bsv
+++ b/CHERICC.bsv
@@ -141,7 +141,7 @@ endinstance
 // CHERICC capability type
 ////////////////////////////////////////////////////////////////////////////////
 
-`define CCSoftPerms Bit#(4)
+`define CCSoftPerms Bit#(2)
 `define AllPermsSz TAdd#(SizeOf#(`CCSoftPerms), SizeOf#(HardPerms))
 
 typedef struct {
@@ -162,7 +162,6 @@ instance Bits#(CHERICCCap#(addr_, bounds_, e_, t_),
     Bit#(1)                         isCap = pack(cap.isCap);
     Bit#(SizeOf#(`CCSoftPerms)) softperms = cap.softperms;
     Bit#(SizeOf#(HardPerms))    hardperms = pack(cap.hardperms);
-    Bit#(res_)                        res = cap.res;
     Bit#(bounds_)                  bounds = pack(cap.bounds);
     Bit#(addr_)                      addr = cap.addr;
     return {isCap, softperms, hardperms, res, bounds, addr};
@@ -290,15 +289,30 @@ instance CHERICap#(CHERICCCap#(addr_, bounds_, e_, t_), t_, addr_) provisos (
     return cap;
   endfunction
   //////////////////////////////////////////////////////////////////////////////
-  function getKind(cap) = case (cap.bounds) matches
+  function getKind(cap) = case (cap.otypes) matches
     tagged Sealed ._: return SEALED_WITH_TYPE;
     default: return UNSEALED;
   endcase;
+  function getColorAwareKind (cap, pidt);
+    if (isCCType(cap, pidt))
+        return COLORED  (cap.otype);
+    else begin
+        case (cap.otype)
+            otype_unsealed: return UNSEALED;
+            otype_sentry:   return SENTRY;
+            otype_res0:     return RES0;
+            otype_res1:     return RES1;
+            default:        return SEALED_WITH_TYPE (cap.otype);
+        endcase
+    end
+  endfunction
+  function setCCTypes(cap, kind);
+    cap.otypes = truncate(kind);
+    return cap;
+  endfunction;
   //////////////////////////////////////////////////////////////////////////////
-  function getType(cap) = case (cap.bounds) matches
-    tagged Sealed .b: return zeroExtend(b.otype);
-    default: return -1;
-  endcase;
+  function getCCType(cap) = zeroExtend(cap.otypes);
+  function isCCType(cap, pidt) = (cap.otype>0 && zeroExtend(cap.otype)<pidt) ? True : False;
   //////////////////////////////////////////////////////////////////////////////
   function setType(cap, otype);
     let new_cap = cap;
diff --git a/CHERICC_Fat.bsv b/CHERICC_Fat.bsv
index 2527c84..90d0696 100644
--- a/CHERICC_Fat.bsv
+++ b/CHERICC_Fat.bsv
@@ -40,6 +40,7 @@ export CapPipe;
 export CapFat;
 export MW;
 export OTypeW;
+export CCTypeW;
 export FlagsW;
 export Perms;
 export ResW;
@@ -86,10 +87,11 @@ typedef `FLAGSW FlagsW;
 typedef 32 CapAddrW;
 typedef 64 CapW;
 `else // CAP128 is default
-typedef 4   UPermW;
+typedef 2   UPermW;
 typedef 14  MW;
 typedef 6   ExpW;
-typedef 18  OTypeW;
+typedef 21  OTypeW;
+typedef 21  CCTypeW;
 typedef `FLAGSW FlagsW;
 typedef 64  CapAddrW;
 typedef 128 CapW;
@@ -114,6 +116,8 @@ typedef Bit#(CapAddrW)          CapAddr;
 typedef Bit#(TAdd#(CapAddrW,1)) CapAddrPlus1;
 typedef Bit#(TAdd#(CapAddrW,2)) CapAddrPlus2;
 // The Hardware permissions type
+
+
 typedef struct {
   Bool permit_set_CID;
   Bool access_sys_regs;
@@ -196,24 +200,44 @@ function Fmt showArchitectural(CapFat cap) =
   + $format(" length:0x%x", getLengthFat(cap, getTempFields(cap)));
 
 // "Microarchitectural FShow"
+// instance FShow#(CapFat);
+//   function Fmt fshow(CapFat cap) =
+//     $format("valid:%b", cap.isCapability)
+//     + $format(" perms:0x%x", getPerms(cap))
+//     //+ $format(" flags:0x%x", getFlags(cap))
+//     + $format(" format:", fshow(cap.format))
+//     + $format(" bounds:", fshow(cap.bounds))
+//     + $format(" address:0x%x", cap.address)
+//     + $format(" addrBits:0x%x", cap.addrBits)
+//     + $format(" {bot:0x%x top:0x%x len:0x%x offset:0x%x}",
+//                 getBotFat(cap, getTempFields(cap)),
+//                 getTopFat(cap, getTempFields(cap)),
+//                 getLengthFat(cap, getTempFields(cap)),
+//                 getOffsetFat(cap, getTempFields(cap)))
+//     + $format(" (TempFields: {") + fshow(getTempFields(cap)) + $format("})");
+// endinstance
 instance FShow#(CapFat);
-  function Fmt fshow(CapFat cap) =
-    $format("valid:%b", cap.isCapability)
-    + $format(" perms:0x%x", getPerms(cap))
-    //+ $format(" flags:0x%x", getFlags(cap))
-    + $format(" reserved:0x%x", cap.reserved)
-    + $format(" format:", fshow(cap.format))
-    + $format(" bounds:", fshow(cap.bounds))
-    + $format(" address:0x%x", cap.address)
-    + $format(" addrBits:0x%x", cap.addrBits)
-    + $format(" {bot:0x%x top:0x%x len:0x%x offset:0x%x}",
-                getBotFat(cap, getTempFields(cap)),
-                getTopFat(cap, getTempFields(cap)),
-                getLengthFat(cap, getTempFields(cap)),
-                getOffsetFat(cap, getTempFields(cap)))
-    + $format(" (TempFields: {") + fshow(getTempFields(cap)) + $format("})");
+  function Fmt fshow(CapFat cap);
+    Fmt str;
+    if (cap.isCapability) 
+      str = $format("[Capability]: ")
+      + $format(" perms:0x%x", getPerms(cap))
+      + $format(" format:", fshow(cap.format))
+      + $format(" bounds:", fshow(cap.bounds))
+      + $format(" address:0x%x", cap.address)
+      + $format(" addrBits:0x%x", cap.addrBits)
+      + $format(" {bot:0x%x top:0x%x len:0x%x offset:0x%x}",
+                  getBotFat(cap, getTempFields(cap)),
+                  getTopFat(cap, getTempFields(cap)),
+                  getLengthFat(cap, getTempFields(cap)),
+                  getOffsetFat(cap, getTempFields(cap)))
+      + $format(" (TempFields: {") + fshow(getTempFields(cap)) + $format("})");
+    else 
+        str = $format("[Data]: ")
+        + $format(" data:0x%x", cap.address);
+    return str;
+  endfunction
 endinstance
-
 // default value for CatFat
 CapFat defaultCapFat = defaultValue;
 
@@ -686,6 +710,9 @@ function CapFat unseal(CapFat cap, x _);
   ret.otype = otype_unsealed;
   return ret;
 endfunction
+function CapFat color(CapFat cap, TempFields tf, CType otype);
+    return seal(cap, tf, otype);
+endfunction
 function VnD#(CapFat) incOffsetFat( CapFat cap
                                   , CapAddr pointer
                                   , CapAddr offset // this is the increment in inc offset, and the offset in set offset
@@ -1085,7 +1112,18 @@ instance CHERICap #(CapMem, OTypeW, FlagsW, CapAddrW, CapW, TSub #(MW, 3));
 
   // capability kind
   //////////////////////////////////////////////////////////////////////////////
+  function getColor = error ("getColor not implemented for CapMem");
+  function getColorAwareKind = error ("getColorAwareKind not implemented for CapMem");
+  function isColoredCap = error ("isColoredCap not implemented for CapMem");
+  function isSealedOrSentry = error ("isSealedOrSentry not implemented for CapMem");
+  function isUnsealedOrColored = error ("isSealedOrSentry not implemented for CapMem");
+  function isKindNot = error ("isSealedOrSentry not implemented for CapMem");
+  function setColor = error ("setColor not implemented for CapMem");
   function getKind = error ("getKind not implemented for CapMem");
+  function getBaseKind = error ("getKind not implemented for CapMem");
+  function getCCType = error ("getCCType not implemented for CapMem");
+  function isCCType = error ("isCCType not implemented for CapMem");
+  function setCCType = error ("setCCType not implemented for CapMem");
   function setKind = error ("setKind not implemented for CapMem");
   function validAsType (dummy, checkType);
     UInt #(CapAddrW) checkTypeUnsigned = unpack (checkType);
@@ -1245,11 +1283,74 @@ instance CHERICap #(CapReg, OTypeW, FlagsW, CapAddrW, CapW, TSub #(MW, 3));
     cap.perms.soft = truncate (perms);
     return cap;
   endfunction
+  function setCCType (cap, otype); 
+    // Only allow changing otype if SW1 permission is set (bit 1 of soft perms)
+    if (cap.perms.soft[1] == 1'b1)
+      cap.otype = truncate (otype);
+    // else: otype remains unchanged (not allowed without SW1)
+    return cap;
+  endfunction
+  function setColor (cap, otype); 
+    cap.otype = truncate (otype);
+    return cap;
+  endfunction
   //function getPerms = error ("getPerms not implemented for CapReg");
   //function setPerms = error ("setPerms not implemented for CapReg");
 
   // capability kind
   //////////////////////////////////////////////////////////////////////////////
+  function getColorAwareKind (cap, ccpt);
+    if (isColoredCap(cap, ccpt))
+         return COLORED (cap.otype);
+    else begin 
+      case (cap.otype)
+        otype_unsealed: return UNSEALED;
+        otype_sentry:   return SENTRY;
+        otype_res0:     return RES0;
+        otype_res1:     return RES1;
+        default:        return SEALED_WITH_TYPE (cap.otype);
+      endcase
+    end
+  endfunction
+  function  Maybe#(Bit#(21)) getColor(cap, ccpt);
+    if (isColoredCap(cap, ccpt))
+         return Valid(cap.otype);
+    else begin 
+      return Invalid;
+    end
+  endfunction
+  // function isColoredCap (cap, ccpt) = (cap.otype>=0 && zeroExtend(cap.otype)<ccpt) ? True : False;
+  function isColoredCap(cap, ccpt);
+    Bool result;
+    if(cap.otype==otype_unsealed)   result = False;
+    else if(cap.otype==otype_sentry) result = False;
+    else if(cap.otype==otype_res0)  result = False;
+    else if(cap.otype==otype_res1)  result = False;
+    else if (zeroExtend(cap.otype)<ccpt) result = True;
+    else result = False;   //otype_sealed
+    return result;
+  endfunction
+
+  function isSealedOrSentry (cap, ccpt);
+    Bool result;
+    if ( isColoredCap(cap, ccpt) ) result=False;
+    else if ( cap.otype == otype_unsealed ) result= False;
+    else if ( cap.otype == otype_res0 ) result= False;
+    else if ( cap.otype == otype_res1 ) result= False;
+    else result= True;
+    return result;
+  endfunction
+  function isUnsealedOrColored (cap, ccpt);
+    let kind = getColorAwareKind(cap,ccpt);
+    return  (kind==UNSEALED) ||
+            (kind matches tagged COLORED .t ? True:False);
+  endfunction
+  function isKindNot (cap, ccpt, kind);
+    Bool result;
+    if ( getColorAwareKind(cap, ccpt) == kind ) result = False;
+    else result = True;
+    return result;
+  endfunction
   function getKind (cap) = case (cap.otype)
     otype_unsealed: UNSEALED;
     otype_sentry:   SENTRY;
@@ -1257,13 +1358,27 @@ instance CHERICap #(CapReg, OTypeW, FlagsW, CapAddrW, CapW, TSub #(MW, 3));
     otype_res1:     RES1;
     default:        SEALED_WITH_TYPE (cap.otype);
   endcase;
+
+  function getBaseKind (cap) = case (cap.otype)
+    otype_unsealed: UNSEALED;
+    otype_sentry:   SENTRY;
+    otype_res0:     RES0;
+    otype_res1:     RES1;
+    default:        SEALED_WITH_TYPE (cap.otype);
+  endcase;
+
+  function getCCType (cap) = zeroExtend (cap.otype);
+  function isCCType (cap, pidt) = (cap.otype>0 && zeroExtend(cap.otype)<pidt) ? True : False;
+  
   function setKind (cap, kind) = case (kind) matches
     tagged UNSEALED:             unseal (cap, ?);
     tagged SENTRY:               seal (cap, ?, VnD {v: True, d:otype_sentry});
     tagged RES0:                 seal (cap, ?, VnD {v: True, d:otype_res0});
     tagged RES1:                 seal (cap, ?, VnD {v: True, d:otype_res1});
+    tagged COLORED .ot:          color (cap, ?, VnD {v: True, d:ot});
     tagged SEALED_WITH_TYPE .ot: seal (cap, ?, VnD {v: True, d:ot});
   endcase;
+
   function validAsType (dummy, checkType);
     CapMem nullC = nullCap;
     return validAsType (nullC, checkType);
@@ -1360,10 +1475,24 @@ instance CHERICap #(CapPipe, OTypeW, FlagsW, CapAddrW, CapW, TSub#(MW, 3));
     CapPipe { capFat: setSoftPerms(cap.capFat, perms)
             , tempFields: cap.tempFields };
   function getKind (cap) = getKind(cap.capFat);
+  function getBaseKind (cap) = getBaseKind(cap.capFat);
+  function getColor (cap) = getColor(cap.capFat);
+  function getColorAwareKind (cap, pidt) = getColorAwareKind(cap.capFat, pidt);
+  function isColoredCap (cap, ccpt) = isColoredCap(cap.capFat, ccpt);
+  function isSealedOrSentry (cap, ccpt) = isSealedOrSentry (cap.capFat, ccpt);
+  function isUnsealedOrColored (cap, ccpt)= isUnsealedOrColored(cap.capFat, ccpt);
+  function isKindNot (cap, ccpt, kind)= isKindNot(cap.capFat, ccpt, kind);
+  function setColor (cap, kind) =
+    CapPipe { capFat:setColor(cap.capFat,kind)
+            , tempFields: cap.tempFields };
+  function getCCType (cap) = getCCType(cap.capFat);
+  function isCCType (cap, pidt) = isCCType(cap.capFat, pidt);
   function setKind (cap, kind) =
     CapPipe { capFat:setKind(cap.capFat,kind)
             , tempFields: cap.tempFields };
-
+  function setCCType (cap, kind) =
+    CapPipe { capFat:setCCType(cap.capFat,kind)
+            , tempFields: cap.tempFields };
   function getMeta (cap) = getMeta (cap.capFat);
   function getAddr (cap) = getAddr (cap.capFat);
   function maskAddr (cap, mask) =
diff --git a/CHERICap.bsv b/CHERICap.bsv
index 2f6d5a4..d9cff49 100644
--- a/CHERICap.bsv
+++ b/CHERICap.bsv
@@ -36,6 +36,7 @@ package CHERICap;
 
 typedef Bit #(16) SoftPerms;
 
+
 typedef struct {
   Bool permitSetCID;
   Bool accessSysRegs;
@@ -69,6 +70,7 @@ endinstance
 
 typedef union tagged {
   void UNSEALED;
+  Bit #(otypeW) COLORED;
   void SENTRY;
   void RES0;
   void RES1;
@@ -195,9 +197,20 @@ typeclass CHERICap #( type capT              // type of the CHERICap capability
   //////////////////////////////////////////////////////////////////////////////
   // Manipulate the kind of the capability, i.e. whether it is sealed, sentry,
   // unsealed, ...
-
+  function capT setColor (capT cap, Bit #(otypeW) kind);
+  function Bool isColoredCap (capT cap, Bit #(otypeW) ccpt);
+  function Bool isSealedOrSentry (capT cap, Bit #(otypeW) ccpt);
+  function Bool isUnsealedOrColored (capT cap, Bit #(otypeW) ccpt);
+  function Bool isKindNot (capT cap, Bit #(otypeW) ccpt, Kind #(otypeW) kind);
+  // get the kind of a capability
+  function Maybe#(Bit#(otypeW)) getColor (capT cap, Bit#(otypeW) ccpt);  
+  function Kind #(otypeW) getColorAwareKind (capT cap, Bit#(otypeW) ccpt);
+  function capT setCCType (capT cap, Bit #(otypeW) kind);
+  function Bit #(32) getCCType (capT cap);
+  function Bool isCCType (capT cap, Bit# (32) pidt);
   // get the kind of a capability
   function Kind #(otypeW) getKind (capT cap);
+  function Kind #(otypeW) getBaseKind (capT cap);
   // set the kind of a capability
   function capT setKind (capT cap, Kind #(otypeW) kind);
   // Check if a type is valid (requires a dummy proxy)
diff --git a/CHERICapWrap.bsv b/CHERICapWrap.bsv
index 0c5402a..2f61092 100644
--- a/CHERICapWrap.bsv
+++ b/CHERICapWrap.bsv
@@ -58,7 +58,11 @@ function Bit#(31) `W(getPerms) (`CAPTYPE cap) = getPerms(cap);
 (* noinline *)
 function `CAPTYPE `W(setPerms) (`CAPTYPE cap, Bit#(31) perms) = setPerms(cap, perms);
 (* noinline *)
-function Kind#(OTypeW) `W(getKind) (`CAPTYPE cap) = getKind(cap);
+function `CAPTYPE `W(setCCType) (`CAPTYPE cap, Bit#(CCTypeW) kind) = setCCType(cap, kind);
+(* noinline *)
+function Bit#(21) `W(getCCType) (`CAPTYPE cap, Bit# (32)) = getCCType(cap, pidt);
+function Bool `W(isCCType) (`CAPTYPE cap) = isCCType(cap);
+function Kind#(CCTypeW) `W(getKind) (`CAPTYPE cap) = getKind(cap);
 (* noinline *)
 function `CAPTYPE `W(setKind) (`CAPTYPE cap, Kind#(OTypeW) kind) = setKind(cap, kind);
 (* noinline *)
diff --git a/CHERI_CAP_API.adoc b/CHERI_CAP_API.adoc
index 3aef3b0..cc378f2 100644
--- a/CHERI_CAP_API.adoc
+++ b/CHERI_CAP_API.adoc
@@ -99,7 +99,7 @@ typedef union {
   void SENTRY;
   void RES0;
   void RES1;
-  Bit #(type_width) SEALED_WITH_TYPE;
+  Bit #(21) SEALED_WITH_TYPE;
 } Kind #(numeric type type_width);
 ----
 
