diff --git a/clang/include/clang/Basic/Builtins.def b/clang/include/clang/Basic/Builtins.def
index 8fb1c1e5ef1e..388f9ff596cd 100644
--- a/clang/include/clang/Basic/Builtins.def
+++ b/clang/include/clang/Basic/Builtins.def
@@ -1654,6 +1654,7 @@ BUILTIN(__builtin_cheri_offset_get, "zvC*m", "nct")
 BUILTIN(__builtin_cheri_offset_increment, "v*mvC*mz", "nct")
 BUILTIN(__builtin_cheri_offset_set, "v*mvC*mY", "nct")
 BUILTIN(__builtin_cheri_perms_and, "v*mvC*mz", "nct")
+BUILTIN(__builtin_cheri_cc_set_type, "v*mvC*mz", "nct")
 BUILTIN(__builtin_cheri_perms_check, "vvC*mCz", "nct")
 BUILTIN(__builtin_cheri_perms_get, "zvC*m", "nct")
 BUILTIN(__builtin_cheri_seal, "v*mvC*mvC*m", "nct")
diff --git a/clang/lib/CodeGen/CGBuiltin.cpp b/clang/lib/CodeGen/CGBuiltin.cpp
index d04a69869abf..e60f70169d15 100644
--- a/clang/lib/CodeGen/CGBuiltin.cpp
+++ b/clang/lib/CodeGen/CGBuiltin.cpp
@@ -4922,6 +4922,14 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
                                 {EmitCastToVoidPtr(Cap), Perms}),
         Cap->getType()));
   }
+  case Builtin::BI__builtin_cheri_cc_set_type: {
+    Value *Cap = EmitScalarExpr(E->getArg(0));
+    Value *Perms = EmitScalarExpr(E->getArg(1));
+    return RValue::get(Builder.CreateBitCast(
+        Builder.CreateIntrinsic(llvm::Intrinsic::cheri_cap_cc_set_type, {SizeTy},
+                                {EmitCastToVoidPtr(Cap), Perms}),
+        Cap->getType()));
+  }
   case Builtin::BI__builtin_cheri_offset_get:
     return RValue::get(Builder.CreateIntrinsic(
         llvm::Intrinsic::cheri_cap_offset_get, {SizeTy},
diff --git a/clang/lib/Sema/SemaChecking.cpp b/clang/lib/Sema/SemaChecking.cpp
index e41e47fdca5b..d873c46988b6 100644
--- a/clang/lib/Sema/SemaChecking.cpp
+++ b/clang/lib/Sema/SemaChecking.cpp
@@ -2507,6 +2507,7 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
   case Builtin::BI__builtin_cheri_bounds_set_exact:
   case Builtin::BI__builtin_cheri_high_set:
   case Builtin::BI__builtin_cheri_perms_and:
+  case Builtin::BI__builtin_cheri_cc_set_type:
   case Builtin::BI__builtin_cheri_flags_set: {
     // The CHERI mutators should accept capability pointer types and
     // uintcap_t and have the same (const-preserving) result type.
diff --git a/llvm/include/llvm/IR/IntrinsicsCHERICap.td b/llvm/include/llvm/IR/IntrinsicsCHERICap.td
index 192930be40ae..d69ddfb3687f 100644
--- a/llvm/include/llvm/IR/IntrinsicsCHERICap.td
+++ b/llvm/include/llvm/IR/IntrinsicsCHERICap.td
@@ -25,6 +25,12 @@ def int_cheri_cap_perms_and :
   Intrinsic<[llvm_cap_ty],
             [llvm_cap_ty, llvm_anyint_ty],
             [IntrNoMem, IntrWillReturn]>;
+
+def int_cheri_cap_cc_set_type :
+  Intrinsic<[llvm_cap_ty],
+            [llvm_cap_ty, llvm_anyint_ty],
+            [IntrNoMem, IntrWillReturn]>;
+
 def int_cheri_cap_perms_get :
   Intrinsic<[llvm_anyint_ty],
             [llvm_cap_ty],
diff --git a/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp b/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp
index 0c53c3f8223e..8a68716ee55b 100644
--- a/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp
+++ b/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp
@@ -1266,6 +1266,7 @@ bool RISCVInstrInfo::isGuaranteedNotToTrap(const llvm::MachineInstr &MI) const {
     // All these instructions were changed to non-trapping.
     switch (MI.getOpcode()) {
     case RISCV::CAndPerm:
+    case RISCV::CCSetType:
     case RISCV::CBuildCap:
     case RISCV::CCopyType:
     case RISCV::CCSeal:
diff --git a/llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td b/llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td
index 63ee0b4f6b09..d12c8024159f 100644
--- a/llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td
+++ b/llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td
@@ -341,6 +341,7 @@ let defsCanBeSealed = 1 in
 def CSeal           : Cheri_rr<0xb, "cseal", GPCR, GPCR>;
 def CUnseal         : Cheri_rr<0xc, "cunseal", GPCR, GPCR>;
 def CAndPerm        : Cheri_rr<0xd, "candperm">;
+def CCSetType        : Cheri_rr<0x15, "ccsettype">;
 } // let mayTrap = 1
 def CSetFlags       : Cheri_rr<0xe, "csetflags">;
 def CSetOffset      : Cheri_rr<0xf, "csetoffset">;
@@ -1204,6 +1205,7 @@ def : PatGpcr<int_cheri_cap_high_get, CGetHigh>;
 def : PatGpcrGpcr<int_cheri_cap_seal, CSeal>;
 def : PatGpcrGpcr<int_cheri_cap_unseal, CUnseal>;
 def : PatGpcrGpr<int_cheri_cap_perms_and, CAndPerm>;
+def : PatGpcrGpr<int_cheri_cap_cc_set_type, CCSetType>;
 def : PatGpcrGpr<int_cheri_cap_flags_set, CSetFlags>;
 def : PatGpcrGpr<int_cheri_cap_offset_set, CSetOffset>;
 def : PatGpcrGpr<int_cheri_cap_address_set, CSetAddr>;
diff --git a/llvm/lib/Target/RISCV/RISCVSystemOperandsXCheri.td b/llvm/lib/Target/RISCV/RISCVSystemOperandsXCheri.td
index f85fbfb1b204..57cc6a565955 100644
--- a/llvm/lib/Target/RISCV/RISCVSystemOperandsXCheri.td
+++ b/llvm/lib/Target/RISCV/RISCVSystemOperandsXCheri.td
@@ -58,6 +58,10 @@ def lookupSpecialCapRegByName : SearchIndex {
 def SCR_PCC : SpecialCapReg<"pcc", 0x00>;
 def SCR_DDC : SpecialCapReg<"ddc", 0x01>;
 
+def CCP : SysReg<"ccp", 0x012>;
+def CCPT : SysReg<"ccpt", 0x013>;
+
+
 //===------------------------------
 // User Trap Setup/Handling
 //===------------------------------
